### PR TITLE
Delegation with extra arguments

### LINKED ISSUE #30: Delegate with additional arguments
Hi! 

I have a use case for delegation with internal mutability with around 100 methods in the same object and `rust-delegate` would be very useful to halve the boilerplate code.

```rust
impl Database {
    delegate! {
        to self.data.borrow_mut() {
            pub fn project(&self, id: &ProjectId) -> Option<Project>;
            // ...
        }
   }
}   
```

However, the method on the inner object always carries an extra argument that should be passed in during delegation:

```rust
pub fn project(&mut self, id: &ProjectId, source: &Source) -> Option<Project> { /* ... */ }
```

It would be useful to have the an ability to append expressions to the argument list for the function called on the inner object, e.g.:

```rust
impl Database {
    delegate! {
        to self.data.borrow_mut() {
            #[extra_args(&self.source)]
            pub fn project(&self, id: &ProjectId) -> Option<Project>;
            // ...
        }
   }
}   
```

### PR DESCRIPTION
**Pull request by kondziu**
Adds an `extra_args` attribute to delegated functions. The `extra_args` attribute takes a comma-separated list of expressions as an argument and appends this list of expressions to the generated function call. Fixes #30.

Note: the list of arguments that was generated so far consisted only of `Ident`s, so the `Ident`s were wrapped into `ExprPath`s to make them type-compatible with the list of `Expr`s passed to `extra_args`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -200,6 +245,19 @@ fn parse_attributes<'a>(
             name = Some(target.name);
         }),
     );
+    map.insert(
+        "extra_args",
+        Box::new(|stream| {
+            let target = syn::parse2::<ExtraArgumentAttribute>(stream).unwrap();
```
Comment (Kobzol) [inline src/lib.rs:251]:
```suggestion
            let target = syn::parse2::<ExtraArgumentAttribute>(stream)
                .expect("extra_args takes a comma-separated list of expressions");
```

**Review by Kobzol — CHANGES_REQUESTED** [review]
Hi! Thanks for your PR. It's seems like a useful use-case indeed.

I wonder if we should extend it to support adding extra args on arbitrary places with placeholders? Something like
```rust
delegate! {
        to self.inner {
            #[extra_args($a = 2)]
            #[extra_args($b = self.default_offset)]
            pub fn method(&self, $a: u32, num: u32, $b: u32) -> u32;
        }
    }
```

Could you please add a simple test for this feature? You can use your code from the README and assert that the delegated method returns the correct values, see existing tests.

**Comment by kondziu** [comment]
Thanks for taking the time to review the PR. Sorry for the late response---I'm supposed to be on vacation and taking time off from coding :)

I will add a test as soon as I'm able (as allowed by the aforementioned vacations). My apologies for missing it in the first place. I must have been very excited about submitting the PR :)

I think adding arguments at arbitrary positions is potentially useful.  I had to rewrite some functions in my code to reverse the argument order in my code, so I even have a horse in the running here :) However, I would worry about some particulars and how they can become preventative though. Some thoughts foryour consideration.

**1**. I would worry that this particular way of annotating it would be confusing. This:

```rust
delegate! {
        to self.inner {
            #[extra_args($a = 2)]
            #[extra_args($b = self.default_offset)]
            pub fn method(&self, $a: u32, num: u32, $b: u32) -> u32;
        }
    }
```

Would generate this:

```rust
pub fn method(&self, num: u32) -> u32 {|
    self.innter.method(2, num, self.default_offset)
}
```

I think the fact that the signature now expresses two function signatures at once might be confusing, especially on first encounter. That is, currently when I see a `delegate` block, it's pretty clear what's going on. If I saw additional arguments, I'd have to read the docs.

**2**. The amount of annotation you write to express arbitrarily placed extra arguments approaches the amount of boiler platecode you'd need to just write the function out verbatim. And the function written out verbatim has the additional advantage of being obvious. I.e. if you need to define the function signature like this:

```rust
delegate! {
        to self.inner {
            #[extra_args($b = 0]
            #[extra_args($d = self.default_offset)]
            pub fn method(&self, a: u32, $b: u32, c: u32, $d: u32) -> u32;
        }
    }
```

You do not gain much, if anything, over writing out the function call verbatim:

```rust
pub fn method(&self, num: u32) -> u32 {|
    self.innter.method(a, 0, num, self.default_offset)
}
```

I'd definitely prefer to see the second option in code that I were reading that somebody else wrote.

**3**. It increases the complexity of the macro a lot. Currently, the argument list is harvested from the signature of the provided function in a very straghforward way.  With this suggestion, a lot of parsing needs to happen. Then, when generating the code 

Given 1 and 2, I wonder if it is worth to do all of 3.

I don't have an equally flexible proposal, but perhaps it would be suffficient to rename the annotation to `append_args` (or similar) and add another one: `prepend_args`  (or similar)? I think it has the advantage of being reasonably intuitive, easy to implement, and has a reasonable coverage of use cases. Thoughts?

**Comment by Kobzol** [comment]
It would introduce a lot of additional complexity in the code, that's for sure. On the other hand, I think that a lot of your arguments actually also apply to your "append-only" arguments. For example
```rust
delegate! {
    to self.inner {
        #[extra_args(0)]
        pub fn method(&self, a: u32) -> u32;
        // vs.
        pub fn method(&self, a: u32) -> u32 {
            self.inner.method(a, 0);
        }
    }
}
```
Doesn't seem like an extra win :D But then again, this crate does not remove a lot of boilerplate that could be removed, because of the limitations of proc macros.

Let's rename `extra_args` to `append_args` and leave the rest as is, to keep it simple for now :) Once you do that and add some tests, I'll merge it.

**Comment by Kobzol** [comment]
Please also apply my previous comment (https://github.com/chancancode/rust-delegate/pull/31#discussion_r649950714). Afterwards I'll merge it and release a new version.

**Comment by kondziu** [comment]
You're the boss, boss.

**Comment by kondziu** [comment]
As a side note, I was thinking about how I'd implement extra arguments at arbitrary positions, and given that your proposition would involve parsing the signature, I'd like to suggest that at that point you might inline the extra argument expressions with some indicator that they are not a part of the signature, e.g.:

```rust
delegate! {
        to self.inner {
            pub fn method(&self, a: u32, [ 0 ], c: u32, [ self.default_offset ]) -> u32;
        }
    }
```

**Comment by Kobzol** [comment]
That actually looks pretty nice! I think that could work, it will just need a little bit of `syn`-fu. I would leave it for a future PR though, let's keep `append_args` for now to solve your use-case.

**Comment by kondziu** [comment]
No objections.

**Comment by Kobzol** [comment]
Thank you very much @kondziu! I'll release a new version soon after this PR is merged.

**Pull request merged**